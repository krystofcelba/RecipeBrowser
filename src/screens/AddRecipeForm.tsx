import React, { Component } from 'react';
import { ScrollView, Image, View, StyleSheet, Text, TextInput } from 'react-native';
import { connect } from 'react-redux';
import { Cell, Section, TableView } from 'react-native-tableview-simple';

import { Colors, StylesConstants } from '../assets/constants';
import i18n from '../assets/i18n';
import FormInputCell from '../components/new-recipe-form/FormInputCell';
import FormButtonCell from '../components/new-recipe-form/FormButtonCell';
import FormIngredientCell from '../components/new-recipe-form/FormIngredientCell';
import { actionCreators, NewRecipe } from '../redux/reducers/ui';

interface Props {
  navigator: any;
  newRecipe: NewRecipe;
  canBeSubmitted: boolean;
  addIngredient: typeof actionCreators.addIngredientToAddRecipeForm;
  updateIngredient: typeof actionCreators.updateIngredientInAddRecipeForm;
  updateName: typeof actionCreators.updateNameInAddRecipeForm;
  updateDescription: typeof actionCreators.updateDescriptionInAddRecipeForm;
}

class AddRecipeForm extends Component<Props> {
  constructor(props) {
    super(props);
    this.props.navigator.setOnNavigatorEvent(this.onNavigatorEvent);
  }

  static navigatorButtons = {
    rightButtons: [
      {
        title: i18n.t('add'), // for a textual button, provide the button title (label)
        id: 'add',
        disabled: true,
      },
    ],
    leftButtons: [
      {
        title: i18n.t('cancel'), // for a textual button, provide the button title (label)
        id: 'cancel',
      },
    ],
  };

  onNavigatorEvent = event => {
    if (event.type === 'NavBarButtonPress') {
      const { navigator } = this.props;
      switch (event.id) {
        case 'cancel':
          navigator.dismissModal();
          break;

        default:
          break;
      }
    }
  };

  render() {
    const {
      newRecipe: { name, description, ingredients },
      updateName,
      updateDescription,
      addIngredient,
      updateIngredient,
    } = this.props;
    return (
      <View style={styles.container}>
        <ScrollView>
          <TableView>
            <Section>
              <FormInputCell placeholder={i18n.t('name')} value={name} onChangeText={updateName} />
              <FormInputCell
                placeholder={i18n.t('description')}
                numberOfLines={4}
                value={description}
                onChangeText={updateDescription}
              />
            </Section>
            <Section header={i18n.t('ingredients')}>
              {Object.keys(ingredients).map(id => {
                const ingredient = ingredients[id];
                return <FormIngredientCell key={id} ingredient={ingredient} onChange={updateIngredient} />;
              })}
              <FormButtonCell title={`+ ${i18n.t('add')}`} onPress={addIngredient} />
            </Section>
          </TableView>
        </ScrollView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#EFEFF4',
    flex: 1,
  },
});

const mapStateToProps = ({ ui: { addRecipeForm: { newRecipe, canBeSubmitted } } }) => ({ newRecipe, canBeSubmitted });

const mapDispatchToProps = {
  updateName: actionCreators.updateNameInAddRecipeForm,
  updateDescription: actionCreators.updateDescriptionInAddRecipeForm,
  addIngredient: actionCreators.addIngredientToAddRecipeForm,
  updateIngredient: actionCreators.updateIngredientInAddRecipeForm,
};

export default connect(mapStateToProps, mapDispatchToProps)(AddRecipeForm);
