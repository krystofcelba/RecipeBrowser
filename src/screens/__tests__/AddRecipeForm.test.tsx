import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import { AddRecipeFormComponent } from '../AddRecipeForm';

describe('AddRecipeForm component', () => {
  const navigator = {
    setOnNavigatorEvent: () => {},
    setButtons: () => {},
  };

  it('renders as expected', () => {
    const setOnNavigatorEvent = jest.fn();
    const setButtons = jest.fn();

    const wrapper = shallow(
      <AddRecipeFormComponent
        navigator={{ ...navigator, setOnNavigatorEvent, setButtons }}
        validationErrors={{ name: 'error', description: 'description error' }}
        invalid={true}
        submit={() => new Promise<any>(resolve => resolve())}
      />,
      {},
    );
    expect(wrapper).toMatchSnapshot();
    expect(setOnNavigatorEvent).toBeCalled();
    expect(setButtons.mock.calls).toMatchSnapshot();
  });

  it('renders valid form as expected', () => {
    const setButtons = jest.fn();

    const wrapper = shallow(
      <AddRecipeFormComponent
        navigator={{ ...navigator, setButtons }}
        validationErrors={{}}
        invalid={false}
        submit={() => new Promise<any>(resolve => resolve())}
      />,
      {},
    );
    expect(wrapper).toMatchSnapshot();
    expect(setButtons.mock.calls).toMatchSnapshot();
  });

  it('calls updateNavigatorButtons on componentWillReceiveProps', () => {
    const wrapper = shallow(
      <AddRecipeFormComponent
        navigator={{ ...navigator }}
        validationErrors={{}}
        invalid={false}
        submit={() => new Promise<any>(resolve => resolve())}
      />,
      {},
    );
    const updateNavigatorButtons = jest.fn();
    wrapper.instance().updateNavigatorButtons = updateNavigatorButtons;
    wrapper.setProps({ navigator, validationErrors: {}, invalid: true });
    expect(updateNavigatorButtons).toBeCalledWith(true);
  });

  it('calls dismissModal on cancel button', () => {
    const dismissModal = jest.fn();

    const wrapper = shallow(
      <AddRecipeFormComponent
        navigator={{ ...navigator, dismissModal }}
        validationErrors={{}}
        invalid={false}
        submit={() => new Promise<any>(resolve => resolve())}
      />,
      {},
    );
    wrapper.instance().onNavigatorEvent({ type: 'NavBarButtonPress', id: 'cancel' });
    expect(dismissModal).toBeCalled();
  });

  it('calls submit on add button', () => {
    const submit = jest.fn();

    const wrapper = shallow(
      <AddRecipeFormComponent navigator={{ ...navigator }} validationErrors={{}} invalid={false} submit={submit} />,
      {},
    );
    wrapper.instance().onNavigatorEvent({ type: 'NavBarButtonPress', id: 'add' });
    expect(submit).toBeCalled();
  });

  it('does not do anything as expected on any other navigator action', () => {
    const wrapper = shallow(
      <AddRecipeFormComponent
        navigator={navigator}
        validationErrors={{}}
        invalid={false}
        submit={() => new Promise<any>(resolve => resolve())}
      />,
    );
    wrapper.instance().onNavigatorEvent({ type: 'anything' });
  });
});
