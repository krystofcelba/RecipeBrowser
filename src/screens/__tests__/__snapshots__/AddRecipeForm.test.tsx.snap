// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRecipeForm component renders as expected 1`] = `
<KeyboardAvoidingView
  style={
    Object {
      "backgroundColor": "#EFEFF4",
      "flex": 1,
    }
  }
  testID="addRecipeFormScreen"
>
  <ScrollViewMock>
    <TableView>
      <Field
        component={[Function]}
        name="image"
      />
      <Section
        allowFontScaling={true}
        footer={null}
        footerComponent={
          <ErrorsSectionFooter
            errors={
              Object {
                "description": "description error",
                "name": "error",
              }
            }
            testIDcontext="base"
          />
        }
        footerTextColor="#6d6d72"
        header={null}
        headerComponent={null}
        headerTextColor="#6D6D72"
        hideSeparator={false}
        sectionPaddingBottom={15}
        sectionPaddingTop={0}
        sectionTintColor="#EFEFF4"
        separatorInsetLeft={15}
        separatorInsetRight={0}
        separatorTintColor="#C8C7CC"
      >
        <TextInputCell
          name="name"
          placeholder="Name"
        />
        <TextInputCell
          name="description"
          numberOfLines={4}
          placeholder="Description"
        />
      </Section>
      <FieldArray
        component={[Function]}
        fieldComponent={[Function]}
        header="Ingredients"
        name="ingredients"
        testIDcontext="ingredients"
      />
      <FieldArray
        component={[Function]}
        fieldComponent={[Function]}
        header="Seasonings"
        name="seasonings"
        placeholder="Seasoning"
        testIDcontext="seasonings"
      />
      <FieldArray
        component={[Function]}
        fieldComponent={[Function]}
        header="Steps"
        name="steps"
        placeholder="Step"
        testIDcontext="steps"
      />
    </TableView>
  </ScrollViewMock>
</KeyboardAvoidingView>
`;

exports[`AddRecipeForm component renders as expected 2`] = `
Array [
  Array [
    Object {
      "leftButtons": Array [
        Object {
          "buttonColor": "#ffff",
          "id": "cancel",
          "testID": "cancel",
          "title": "Cancel",
        },
      ],
      "rightButtons": Array [
        Object {
          "buttonColor": "#ffff",
          "disabled": true,
          "id": "add",
          "testID": "add",
          "title": "Add",
        },
      ],
    },
  ],
]
`;

exports[`AddRecipeForm component renders valid form as expected 1`] = `
<KeyboardAvoidingView
  style={
    Object {
      "backgroundColor": "#EFEFF4",
      "flex": 1,
    }
  }
  testID="addRecipeFormScreen"
>
  <ScrollViewMock>
    <TableView>
      <Field
        component={[Function]}
        name="image"
      />
      <Section
        allowFontScaling={true}
        footer={null}
        footerComponent={
          <ErrorsSectionFooter
            errors={
              Object {
                "description": undefined,
                "name": undefined,
              }
            }
            testIDcontext="base"
          />
        }
        footerTextColor="#6d6d72"
        header={null}
        headerComponent={null}
        headerTextColor="#6D6D72"
        hideSeparator={false}
        sectionPaddingBottom={15}
        sectionPaddingTop={0}
        sectionTintColor="#EFEFF4"
        separatorInsetLeft={15}
        separatorInsetRight={0}
        separatorTintColor="#C8C7CC"
      >
        <TextInputCell
          name="name"
          placeholder="Name"
        />
        <TextInputCell
          name="description"
          numberOfLines={4}
          placeholder="Description"
        />
      </Section>
      <FieldArray
        component={[Function]}
        fieldComponent={[Function]}
        header="Ingredients"
        name="ingredients"
        testIDcontext="ingredients"
      />
      <FieldArray
        component={[Function]}
        fieldComponent={[Function]}
        header="Seasonings"
        name="seasonings"
        placeholder="Seasoning"
        testIDcontext="seasonings"
      />
      <FieldArray
        component={[Function]}
        fieldComponent={[Function]}
        header="Steps"
        name="steps"
        placeholder="Step"
        testIDcontext="steps"
      />
    </TableView>
  </ScrollViewMock>
</KeyboardAvoidingView>
`;

exports[`AddRecipeForm component renders valid form as expected 2`] = `
Array [
  Array [
    Object {
      "leftButtons": Array [
        Object {
          "buttonColor": "#ffff",
          "id": "cancel",
          "testID": "cancel",
          "title": "Cancel",
        },
      ],
      "rightButtons": Array [
        Object {
          "buttonColor": "#ffff",
          "disabled": false,
          "id": "add",
          "testID": "add",
          "title": "Add",
        },
      ],
    },
  ],
]
`;
